### ISSUE TITLE
Signals do not appear to be working

### ISSUE THREAD

**Issue by AndrewGuenther** [jrobichaud/django-structlog#75]
I'm trying to add some additional request and tracing information via the `bind_extra_request_metadata` signal, but this does not appear to be working. Here's a basic handler that can reproduce the issue:

```python
@receiver(signals.bind_extra_request_metadata)
def bind_additional_values(request, logger, **kwargs):
    print("hello")

    logger.bind(test="hello")
```

I can see the print occurring, but the added key does not appear in the logs. I don't actually understand how this ever would have worked because `bind()` returns a new logger with the bound values, it doesn't actually modify `logger`. Similarly, I can see that none of the bound values attached in the middleware are logged for the same reason:

https://github.com/jrobichaud/django-structlog/blob/master/django_structlog/middlewares/request.py#L45
https://github.com/jrobichaud/django-structlog/blob/master/django_structlog/middlewares/request.py#L51

Neither request_id nor ip are included in the logs. Was there some older version of structlog where this may have worked? I'm currently using structlog `21.5.0`

**Comment by jrobichaud** [comment]
Is your file imported?

**Comment by AndrewGuenther** [comment]
Yes, I can see that the signal is executing, but no calls to `bind()` actually mutate the logger, neither in my own signal handler nor the ones I linked that are internal to `django_structlog`

**Comment by jrobichaud** [comment]
Ate you able to reproduce the problem in the demo project?

**Comment by jrobichaud** [comment]
https://django-structlog.readthedocs.io/en/latest/development.html#start-demo-app

**Comment by AndrewGuenther** [comment]
Some relevant docs from stuctlog: https://www.structlog.org/en/stable/thread-local.html#downsides-caveats

**Comment by AndrewGuenther** [comment]
It works in the demo app. Let me try setting the Django and structlog versions in the demo app to more recent versions and see if it still works.

**Comment by AndrewGuenther** [comment]
Basically everything blows up when trying to upgrade to Django 3. It'll take me a while to bring this all up to date.

**Comment by jrobichaud** [comment]
You should not have to upgrade django.

**Comment by AndrewGuenther** [comment]
I'm just trying to get it as close to my environment as possible in order to reproduce the issue, but I'll give it a shot only updating structlog.

**Comment by AndrewGuenther** [comment]
Still works with latest structlog. But I'm still not sure how? Calling `bind()` doesn't actually mutate the logger, the current logic would just be throwing the returned logger away, right?

**Comment by jrobichaud** [comment]
```python3
structlog.configure(
    processors=[
        structlog.stdlib.filter_by_level,
        structlog.processors.TimeStamper(fmt="iso"),
        structlog.stdlib.add_logger_name,
        structlog.stdlib.add_log_level,
        structlog.stdlib.PositionalArgumentsFormatter(),
        structlog.processors.StackInfoRenderer(),
        structlog.processors.format_exc_info,
        structlog.processors.UnicodeDecoder(),
        structlog.stdlib.ProcessorFormatter.wrap_for_formatter,
    ],
    context_class=structlog.threadlocal.wrap_dict(dict),   # <------- maybe this?
    logger_factory=structlog.stdlib.LoggerFactory(),
    wrapper_class=structlog.stdlib.BoundLogger,   # <------- maybe this?
    cache_logger_on_first_use=True,
)
```

**Comment by AndrewGuenther** [comment]
```python
context_class=structlog.threadlocal.wrap_dict(dict), 
```

This was the one! I'm curious why that works though? The docs seem to hammer the point that you need to be saving the result of `bind()` why does this make that not necessary? What are the side effects?

**Comment by jrobichaud** [comment]
`structlog` is intended to be used with python in general and wants you to avoid some pitfalls of randomly forgetting of renewing context.
`django-structlog` is intended to be used with django. How it is implemented the context is renewed with each new request and it is just fine.

**Comment by jrobichaud** [comment]
In future versions we could work to follow all structlog recommendations but it will complexity how django-structlog works.

**Comment by jrobichaud** [comment]
I think you could close this issue. But make sure to open new ones if you find any problems with the project.

**Comment by AndrewGuenther** [comment]
Ah, I'm starting to understand. By wrapping the context with `threadlocal.wrap_dict()` it's effectively saying "any time I add something to the context, do it for everything on this thread" instead of the default behavior of "bind to the returned logger" is that right?

Part of the reason I want to understand is that we share a structlog config with some non-Django projects as well so I want to make sure I don't introduce any odd behaviors there. Seems like we'd want setting `context_class` to be a Django only thing.

Thanks for all of your help on this!

**Comment by jrobichaud** [comment]
> Ah, I'm starting to understand. By wrapping the context with threadlocal.wrap_dict() it's effectively saying "any time I add something to the context, do it for everything on this thread" instead of the default behavior of "bind to the returned logger" is that right?

This is the magic that keeps django-structlog easy to use.

Your project's context and your experience could help to improve how django-structlog works or add complementary documentation to help for other projects like yours.

**Comment by AndrewGuenther** [comment]
In our other projects, we just use `structlog.threadlocal.bind_threadlocal()` for use cases like this. Is there a reason that won't work in `django-structlog`?

**Comment by jrobichaud** [comment]
django-structlog was implemented in mind of not passing around a logger, it would require some refactors in order to make it work without the context_class and keep backward compatibility for most of other users.

**Comment by jrobichaud** [comment]
Could you use a different config for django and  your other project? And the django code using your share code could pass as parameter the logger?

**Comment by AndrewGuenther** [comment]
You don't have to pass around a logger to use `structlog.threadlocal.bind_threadlocal()`, it will apply to all loggers on the thread, isn't that exactly what you want? This is how structlog suggests doing things in their Flask example: https://www.structlog.org/en/stable/examples.html#flask-and-thread-local-data we use this model to attach request ids in all our non-Django services and haven't had any problems.

We can totally have a separate config for our Django app, I'm not worried about that. I'm just trying to suggest a way you could simplify django-structlog's internals.

**Comment by jrobichaud** [comment]
Maybe

**Comment by jrobichaud** [comment]
Yeah but in order to have the bound logger to work you need to have access to that variable in the code of the views.

**Comment by AndrewGuenther** [comment]
You don't. `bind_threadlocal()` applies to all loggers. Here's an example:

```python
import logging

import structlog

structlog.configure(
    processors=[
        structlog.threadlocal.merge_threadlocal,    # Must include this
        structlog.processors.add_log_level,
        structlog.processors.StackInfoRenderer(),
        structlog.dev.set_exc_info,
        structlog.processors.TimeStamper(),
        structlog.dev.ConsoleRenderer()
    ],
    wrapper_class=structlog.make_filtering_bound_logger(logging.NOTSET),
    context_class=dict,
    logger_factory=structlog.PrintLoggerFactory(),
    cache_logger_on_first_use=False
)

logger = structlog.get_logger()
logger.info("Test")

structlog.threadlocal.bind_threadlocal(test="Hello, world")

logger.info("Test 2")
```

Output:

```
1641520011.352684 [info     ] Test
1641520011.3527257 [info     ] Test 2                         test=Hello, world
```

Using `bind_threadlocal()` will update all loggers in the current thread. No need to pass anything around.

**Comment by jrobichaud** [comment]
Ok, I will have to experiment that. How will that clear the loggers on a new request?

**Comment by AndrewGuenther** [comment]
You could call `structlog.threadlocal.clear_threadlocal()` at the end of the request which would clear out all threadlocal data, whether it is owned by `django-structlog` or not, which is what the structlog docs do in their Flask example, or you can unset specific fields at the end of the middleware via `structlog.threadlocal.unbind_threadlocal()` which takes one or more keys.

I'm happy to submit a PR testing it out. I haven't taken a look at the test suite just yet, is it comprehensive enough to give confidence in a change like this? Or should some additional tests be added?

**Comment by AndrewGuenther** [comment]
Absolutely. I suspect the killer here will be that the `structlog.threadlocal.merge_threadlocal` processor would now be required which I don't believe it was before. I'll check that first thing tomorrow. Signing off for the evening.

**Comment by AndrewGuenther** [comment]
Tested it out. The change definitely isn't backwards compatible, but I'd strongly suggest it is worth doing as a major version bump. The current guidance of this package leads to a lot of unexpected behavior by making every `bind()` call attach to a threadlocal context when that isn't expected. As an example:

```python
import logging

import structlog

structlog.configure(
    processors=[
        structlog.threadlocal.merge_threadlocal,    # Must include this
        structlog.processors.add_log_level,
        structlog.processors.StackInfoRenderer(),
        structlog.dev.set_exc_info,
        structlog.processors.TimeStamper(),
        structlog.dev.ConsoleRenderer()
    ],
    wrapper_class=structlog.make_filtering_bound_logger(logging.NOTSET),
    context_class=structlog.threadlocal.wrap_dict(dict),
    logger_factory=structlog.PrintLoggerFactory(),
    cache_logger_on_first_use=False
)

logger_a = structlog.get_logger("a")
logger_b = structlog.get_logger("b")

logger_a.bind(logger_a_key="Hello")
logger_a.info("Test")

logger_b.info("Test 2")
```

Output:
```
1641581575.9166899 [info     ] Test                           logger_a_key=Hello
1641581575.9167335 [info     ] Test 2                         logger_a_key=Hello
```

`logger_a` pollutes `logger_b`'s context in a way that it shouldn't because the required configuration to make `django-structlog` work forces all loggers to share a global context and changes the expected behavior of `bind()`.

**Comment by jrobichaud** [comment]
`django-structlog` was intended to be able to track any logs based on the request's data and be very simple to use.

For existing project using python loggers they just replace with structlog logger and it just works.

If it gets more complicated than that people will not use it.

For this reason I believe we should keep the current behaviour as default and add the support of the more standard behaviour as optional.

**Comment by jrobichaud** [comment]
However the required changes to be able to support the standard behaviour have to be done.

**Comment by jrobichaud** [comment]
I created a branch based on master `improve_structlog_bindings`. If you make a pull request, use it as a base branch please.

**Comment by AndrewGuenther** [comment]
What I'm suggesting however doesn't make it any harder to use. If anything I'd argue it makes it easier because it would still have all the benefits you cited, but would also follow structlog's expected behavior. What about this change would make `django-structlog` more difficult to use?

**Comment by jrobichaud** [comment]
If the breaking change is:
- use `structlog.threadlocal.bind_threadlocal(test="Hello, world")` instead of `logger.bind()` by django-structlog
- remove `context_class=structlog.threadlocal.wrap_dict(dict),` from configuration
- update how signals work

It is still fine for me.

**Comment by jrobichaud** [comment]
An example of change I would not like:
- users would have to change every logger imports they have. Ex: wrap structlog logger with some sort of function 

But I think this is not required.

**Comment by AndrewGuenther** [comment]
That wouldn't be required. The only user-facing change would be to swap `context_class=structlog.threadlocal.wrap_dict(dict)` for adding the `structlog.threadlocal.merge_threadlocal` processor to their configuration (if it isn't already there). The resulting behavior would be the same as long as users are not depending on the divergent `bind()` behavior. If they are, they'd need to switch to using `structlog.threadlocal.bind_threadlocal()` which is the standard behavior.

**Comment by jrobichaud** [comment]
It is still manageable.

**Comment by AndrewGuenther** [comment]
Re-using my example above:

```python
import logging

import structlog

structlog.configure(
    processors=[
      # structlog.threadlocal.merge_threadlocal,
        structlog.processors.add_log_level,
        structlog.processors.StackInfoRenderer(),
        structlog.dev.set_exc_info,
        structlog.processors.TimeStamper(),
        structlog.dev.ConsoleRenderer()
    ],
    wrapper_class=structlog.make_filtering_bound_logger(logging.NOTSET),
    context_class=structlog.threadlocal.wrap_dict(dict),
    logger_factory=structlog.PrintLoggerFactory(),
    cache_logger_on_first_use=False
)

logger_a = structlog.get_logger("a")
logger_b = structlog.get_logger("b")

logger_a.bind(logger_a_key="Hello")
logger_a.info("Test")

logger_b.info("Test 2")
```

Output:
```
1641581575.9166899 [info     ] Test                           logger_a_key=Hello
1641581575.9167335 [info     ] Test 2                         logger_a_key=Hello
```

If a consumer is depending on the behavior that calling `bind()` on `logger_a` would also manipulate `logger_b`, they would need to update the code to:

```python
import logging

import structlog

structlog.configure(
    processors=[
        structlog.threadlocal.merge_threadlocal,
        structlog.processors.add_log_level,
        structlog.processors.StackInfoRenderer(),
        structlog.dev.set_exc_info,
        structlog.processors.TimeStamper(),
        structlog.dev.ConsoleRenderer()
    ],
    wrapper_class=structlog.make_filtering_bound_logger(logging.NOTSET),
  # context_class=structlog.threadlocal.wrap_dict(dict),
    logger_factory=structlog.PrintLoggerFactory(),
    cache_logger_on_first_use=False
)

logger_a = structlog.get_logger("a")
logger_b = structlog.get_logger("b")

structlog.threadlocal.bind_threadlocal(logger_a_key="Hello")
logger_a.info("Test")

logger_b.info("Test 2")
```

in order to get the same output

**Comment by jrobichaud** [comment]
Marvelous! There is a section in the documentation about breaking changes.

I am not sure if you are aware but there is a also a celery integration that passes the bound data to celery's task in order to be bound by the task.

We will have to use `get_threadlocal()` in order to get the bound values and then replace the bind with `structlog.threadlocal.bind_threadlocal(...the_data)`

**Comment by jrobichaud** [comment]
We should maybe consider #64 while fixing this.

**Comment by AndrewGuenther** [comment]
I'll look at submitting a PR some time next week for this

**Comment by AndrewGuenther** [comment]
Started on this today and included #64. Got all the request middleware tests working, but still need to tweak the celery tests.

**Comment by jrobichaud** [comment]
Feel free to open a draft PR to the branch I made.
